ckanext/canada/harvesters.py

```diff
@@ -0,0 +1,616 @@
+            'title': 'Portal Sync',
+            'title_translated': {
+                'en': 'Portal Sync',
+                'fr': 'FR Portal Sync FR',
+            },
+            'description': 'Syncs Datasets, Resources, Views, and DataStores from the Registry to the Portal.',
+            'description_translated': {
+                'en': 'Syncs Datasets, Resources, Views, and DataStores from the Registry to the Portal.',
+                'fr': 'FR Syncs Datasets, Resources, Views, and DataStores from the Registry to the Portal. FR',
+            },
+            'form_config_interface': 'Text'
+        }
+
+
+    def validate_config(self, config):
+        return {}
+
+
+    def gather_stage(self, harvest_job):
+        '''
+        The gather stage will receive a HarvestJob object and will be
+        responsible for:
+            - gathering all the necessary objects to fetch on a later.
+              stage (e.g. for a CSW server, perform a GetRecords request)
+            - creating the necessary HarvestObjects in the database, specifying
+              the guid and a reference to its job. The HarvestObjects need a
+              reference date with the last modified date for the resource, this
+              may need to be set in a different stage depending on the type of
+              source.
+            - creating and storing any suitable HarvestGatherErrors that may
+              occur.
+            - returning a list with all the ids of the created HarvestObjects.
+            - to abort the harvest, create a HarvestGatherError and raise an
+              exception. Any created HarvestObjects will be deleted.
+
+        NOTE:
+        PortalSync: Gathers packages via changed_packages_activity_timestamp_since.
+                    Saves HarvestObjects with the content of the Source Package, Resource Views, and DataDictionaries.
+
+        :param harvest_job: HarvestJob object
+        :returns: A list of HarvestObject ids
+        '''
+        log.info('In PortalSync gather_stage (%s)', harvest_job.source.url)
+
+        # Request only the activity since last successful run
+        last_error_free_job = self.last_error_free_job(harvest_job)
+        log.info('Last error-free job: %r', last_error_free_job)
+        last_time = last_error_free_job.gather_started
+
+        # Note: SOLR works in UTC, and gather_started is also UTC, so
+        # this should work as long as local and remote clocks are
+        # relatively accurate. Going back a little earlier, just in case.
+        get_changes_since = (last_time - datetime.timedelta(hours=1)).isoformat()
+        log.info('Searching for activity since: %s UTC', get_changes_since)
+
+        data = self.registry.action.changed_packages_activity_timestamp_since(since_time=last_time.isoformat())
+        if not data:
+            # There is no activity since last successful run
+            return []
+
+        registry_packages = []
+
+        for result in data:
```

Comment:
the same package id will often appear many times in a row when updates are done, better to get a set of package ids to eliminate duplicate work like the old script does